### PR TITLE
[Feature] Implement option to install Gazebo from source.

### PR DESCRIPTION
**Pull request by rosatamsen**
- [X] Create basic Dockerfile that builds dependencies and sets up source repos for Gazebo 9
- [x] Add shared volume between container and host file system for the source code directory
- [x] Add `--source` option to gzdev command
- [x] Support Gazebo 7, 8, and 9 source install utilizing a more generic Dockerfile as template

This change is [Reviewable](https://reviewable.io/reviews/osrf/gzdev/4)

### CODE REVIEW

File: docker/gazebo/gzsource/Dockerfile [hunk 1]
```diff
@@ -17,8 +17,18 @@ RUN bash -c 'source /etc/os-release && \
 RUN curl https://xpra.org/gpg.asc | apt-key add -
 RUN curl http://packages.osrfoundation.org/gazebo.key | apt-key add -
 RUN apt-get update && apt-get install -y xpra=2.3.*
-RUN apt-get update && apt-get install -y gazebo$GZV
-RUN apt-get clean && rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*
+RUN apt-get update && apt-get install -y build-essential cmake debhelper mesa-utils cppcheck xsltproc python-lxml python-psutil python bc netcat-openbsd gnupg2 net-tools locales
+RUN apt-get update && apt-get install -y libfreeimage-dev libprotoc-dev libprotobuf-dev protobuf-compiler freeglut3-dev libcurl4-openssl-dev libtinyxml-dev libtar-dev libtbb-dev libogre-1.9-dev libxml2-dev pkg-config qtbase5-dev libqwt-qt5-dev libltdl-dev libgts-dev libboost-thread-dev libboost-signals-dev libboost-system-dev libboost-filesystem-dev libboost-program-options-dev libboost-regex-dev libboost-iostreams-dev libbullet-dev libsimbody-dev libignition-common-dev libignition-fuel-tools-dev libignition-transport4-dev libignition-math4-dev libignition-msgs-dev libtinyxml2-dev libsdformat6-dev
```
Comment (j-rivero) [inline docker/gazebo/gzsource/Dockerfile]: these are probably the dependencies for gazebo9. If we hardcoded them here directly, how are we going to handle the different versions of Gazebo supported?

Also please note that the .repos file will bring the sources for ignition, sdformat and gazebo so we probably don't want to have the binary packages installed.

Comment (j-rivero) [reply docker/gazebo/gzsource/Dockerfile]: > these are probably the dependencies for gazebo9. If we hardcoded them here directly, how are we going to handle the different versions of Gazebo supported?

Ouch just read the TODO item list. The second comment of binary packages is still valid though.

Comment (rosatamsen) [reply docker/gazebo/gzsource/Dockerfile]: Yes, I had noticed that too but I wanted to compile only Gazebo first to make sure that works and then remove the binary packages and compile with the sources of all the other dependencies.

**Comment by j-rivero** [comment]
I run this branch using:

```
jrivero@nium gzdev $ (gz-src-install) ./gzdev.py spawn --gzv=9 --source --nvidia

~~~ Spawning docker container for Gazebo 9 ~~~

-> Building docker image. The first time will take a few minutes... 
   You might want to grab a cup of coffee or whatever suits your cup of tea :)

-> Running docker container and forwarding hardware accelerated graphics to your screen

-> Logging output and errors to "/home/jrivero/code/gzdev/gz9_src.log".

jrivero@nium gzdev $ (gz-src-install) cat /home/jrivero/code/gzdev/gz9_src.log
~~~ Container Log ~~~

Summary: 0 packages finished [0.60s]
/bin/bash: /tmp/gazebo/install/local_setup.bash: No such file or directory

~~~ Client log ~~~

NVIDIA Docker: 1.0.1

Client:
 Version:           18.06.0-ce
 API version:       1.38
 Go version:        go1.10.3
 Git commit:        0ffa825
 Built:             Wed Jul 18 19:11:02 2018
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          18.06.0-ce
  API version:      1.38 (minimum version 1.12)
  Go version:       go1.10.3
  Git commit:       0ffa825
  Built:            Wed Jul 18 19:09:05 2018
  OS/Arch:          linux/amd64
  Experimental:     false
8a13be7c3ef3ee637e03364720984c59a09e532827cbc10b639a8753d5731739
Succesfully stopped and removed running container.
```
I'm not sure which is the expected result but if there is no bug, we should point the user about what to do. My expectations were to land inside the docker container with the sources at hand :)

**Comment by j-rivero** [comment]
I got several errors when calling gzcolcon.sh:
```
[19.612s] ERROR:colcon.colcon_core.package_identification:Exception in package identification extension 'python_setup_py' in '/usr/lib/python2.7/dist-packages/numpy/core': No module named 'numpy'
```
If I install python3-numpy in the container:

```
[18.995s] ERROR:colcon.colcon_core.package_identification:Exception in package identification extension 'python_setup_py' in '/usr/lib/python2.7/dist-packages/numpy': setup() function invoked without the keyword argument 'name'
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/colcon_core/package_identification/__init__.py", line 135, in _identify
    retval = extension.identify(desc2)
  File "/usr/lib/python3/dist-packages/colcon_python_setup_py/package_identification/python_setup_py.py", line 46, in identify
    data = extract_data(**kwargs)
  File "/usr/lib/python3/dist-packages/colcon_python_setup_py/package_identification/python_setup_py.py", line 162, in extract_data
    "setup() function invoked without the keyword argument 'name'")
RuntimeError: setup() function invoked without the keyword argument 'name'

[18.999s] ERROR:colcon.colcon_core.package_identification:Exception in package identification extension 'python_setup_py' in '/usr/lib/python2.7/dist-packages/numpy/compat': setup() function invoked without the keyword argument 'name'
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/colcon_core/package_identification/__init__.py", line 135, in _identify
    retval = extension.identify(desc2)
  File "/usr/lib/python3/dist-packages/colcon_python_setup_py/package_identification/python_setup_py.py", line 46, in identify
    data = extract_data(**kwargs)
  File "/usr/lib/python3/dist-packages/colcon_python_setup_py/package_identification/python_setup_py.py", line 162, in extract_data
    "setup() function invoked without the keyword argument 'name'")
RuntimeError: setup() function invoked without the keyword argument 'name'

[19.127s] ERROR:colcon.colcon_core.package_identification:Exception in package identification extension 'python_setup_py' in '/usr/lib/python2.7/dist-packages/numpy/core': No module named 'numpy._build_utils'
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/colcon_core/package_identification/__init__.py", line 135, in _identify
    retval = extension.identify(desc2)
  File "/usr/lib/python3/dist-packages/colcon_python_setup_py/package_identification/python_setup_py.py", line 45, in identify
    kwargs = get_setup_arguments(setup_py)
  File "/usr/lib/python3/dist-packages/colcon_python_setup_py/package_identification/python_setup_py.py", line 108, in get_setup_arguments
    runpy.run_path(str(setup_py))
  File "/usr/lib/python3.5/runpy.py", line 254, in run_path
    pkg_name=pkg_name, script_name=fname)
  File "/usr/lib/python3.5/runpy.py", line 96, in _run_module_code
    mod_name, mod_spec, pkg_name, script_name)
  File "/usr/lib/python3.5/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/usr/lib/python2.7/dist-packages/numpy/core/setup.py", line 13, in <module>
    from numpy._build_utils.apple_accelerate import (uses_accelerate_framework,
ImportError: No module named 'numpy._build_utils'
```
Seems like it is not a fatal error and colcon is able to continue. The next error I found is:

```
removing '/build/complex-dist/build/bdist.linux-x86_64/egg' (and everything under it)
--- stderr: FortranCInterface
CMake Error at CMakeLists.txt:13 (project):
  No CMAKE_Fortran_COMPILER could be found.

  Tell CMake where to find the compiler by setting either the environment
  variable "FC" or the CMake cache entry CMAKE_Fortran_COMPILER to the full
  path to the compiler, or to the compiler name if it is in the PATH.


---
Failed   <<< FortranCInterface	[ Exited with code 1 ]
--- stderr: IntelFortranImplicit
CMake Error at CMakeLists.txt:2 (project):
  No CMAKE_Fortran_COMPILER could be found.

  Tell CMake where to find the compiler by setting either the environment
  variable "FC" or the CMake cache entry CMAKE_Fortran_COMPILER to the full
  path to the compiler, or to the compiler name if it is in the PATH.


---
Failed   <<< IntelFortranImplicit	[ Exited with code 1 ]
```
And the build fails.

A part from these errors (needs fixing) the rest seems to works fine. I would do the gzsources.sh step automatically and leave only the compilation step ready for the user.

**Comment by rosatamsen** [comment]
Hmm could you please double check that the latest changes were actually pulled? Also check if the image was built correctly, the size of the gz9_src image should be 1.71GB.

After running `gzdev spawn 9 --source` you should get the following instructions:

```
Access the container with the following command:
     docker exec -it gz9_src /bin/bash
Once in the container:
     - Pull all source code repos with: gzrepos.sh
     - Then compile and run with verbose output using: gzcolcon.sh
       or less output messages with: colcon build && gazebo
```

and everything should work fine with those commands since they are running in the container and all the prerequistes have been installed already.

**Comment by j-rivero** [comment]
> and everything should work fine with those commands since they are running in the container and all the prerequistes have been installed already.

You are right, the thing works fine after removing the image and launch the command again.

**Comment by j-rivero** [comment]
Merged! good work David!

**Pull request merged**